<!--
This launch file written by both Svena and Simon, contains launch instructions for both leader and follower robot
This is intended as the MAIN launch file. Run as 'roslaunch smartcarts my_launch.launch'

Robot Follower has the following functions:
- Target Tracking (target_pose_tracking.py) & Tf Broadcasting (target_pose_broadcaster.py)
- Data Logger (ball_waypoint_listener.py)
- Motion Controller (follower_controller.py)
- Visualization (trajectory - hector_trajectory_server, tf frames - robot_pose_ekf)

Robot Leader has the following functions:
- Motion Controller (leader_controller.py)
- Visualization (trajectory - hector_trajectory_server)

Global Nodes:
- Waypoints (waypoint_memory.py)
- Simulation Visualization (rviz)
-->

<launch>
  <arg name="waypoints_file" value="$(find smartcarts)/node/preset_poses/sine_L4A1.csv"/> <!-- load in preset poses for leader -->
  <param name="/camera_type" value="depth"/>
  <param name="/ball_radius" value="0.04" />
  <param name="/debug_mode" value="0"/>
  <param name="min_follow_distance" value="0.7"/>
  <param name="min_camera_distance" value="0.7"/>

  <include file="$(find gazebo_ros)/launch/empty_world.launch">
      <env name="GAZEBO_RESOURCE_PATH" value="$(find smartcarts)"/>
      <env name="GAZEBO_MODEL_PATH" value="$(find smartcarts)/models"/>
      <arg name="world_name" value="$(find smartcarts)/worlds/smartcarts.world"/>
      <arg name="gui" value="false"/>
  </include>

  <group ns="follower">
    <include file="$(find smartcarts)/launch/robot_follower.launch">
      <arg name="init_pose" value="-x -1.0 -y 0.0 -z 1.0 -R 0.0 -P 0.0 -Y 0.0" />
      <arg name="robot_name" value="follower" />
    </include>
    <node pkg="smartcarts" type="target_pose_tracking.py" name="follower_ball_tracker"
        output="screen"/>
    <!--TF Broadcasting of Derived Leader's Waypoint Tracking-->
    <node pkg="smartcarts" type="target_pose_broadcaster.py" name="waypoint_broadcaster"
        output="screen"/>
    <node pkg="smartcarts" type="ball_waypoint_listener.py" name="waypoint_listener"
        output="screen"/>

    <node pkg="smartcarts" type="follower_controller.py" name="follower_controller" output="screen"/>

    <node pkg="hector_trajectory_server" type="hector_trajectory_server" name="hts_gazebo_ws" output="screen">
      <param name="target_frame_name" type="string" value="/odom" />
      <param name="source_frame_name" type="string" value="/follower" />
      <param name="trajectory_update_rate" type="double" value="10" />
      <param name="trajectory_publish_rate" type="double" value="5" />
    </node>

    <node pkg="robot_pose_ekf" type="robot_pose_ekf" name="robot_pose_ekf">
      <param name="output_frame" value="odom"/>
      <param name="freq" value="30.0"/>
      <param name="sensor_timeout" value="1.0"/>
      <param name="odom_used" value="true"/>
      <param name="imu_used" value="true"/>
      <param name="vo_used" value="false"/>
      <param name="debug" value="false"/>
      <param name="self_diagnose" value="false"/>
    </node>
  </group>

  <group ns="leader">
    <include file="$(find smartcarts)/launch/robot_leader.launch">
      <arg name="init_pose" value="-x 0.0 -y 0.0 -z 1.0 -R 0.0 -P 0.0 -Y 0.0" />
      <arg name="robot_name" value="leader" />
    </include>
    <node pkg="smartcarts" type="leader_controller.py" name="my_leader" output="screen">
      <param name="waypoints_file" value="$(arg waypoints_file)"/>
    </node>
    <!-- <node pkg="smartcarts" type="line_follow.py" name="leader_line_follow" output="screen"/> -->
    <node pkg="hector_trajectory_server" type="hector_trajectory_server" name="hector_trajectory_server" output="screen">
      <param name="target_frame_name" type="string" value="/odom" />
      <param name="source_frame_name" type="string" value="/leader" />
      <param name="trajectory_update_rate" type="double" value="10" />
      <param name="trajectory_publish_rate" type="double" value="5" />
    </node>
  </group>

  <node pkg="smartcarts" type="waypoint_memory.py" name="waypoint_memory" output="screen"/>

  <node pkg="rviz" type="rviz" name="rviz" args="-d $(find smartcarts)/rviz/follower_pathview.rviz" />

</launch>
